### ISSUE TITLE
Never stop retrying

### ISSUE THREAD

**Issue by kkaefer** [mapbox/mapbox-gl-native#2728]
We're currently stopping retries after 3 attempts. For some types of errors, we are using an exponential timeout until the next retry.

instead, we should *never* stop retrying, but just keep the exponential timeout. The idea is that the implementing application does the right thing and not point to any invalid sources, but that all errors are due to some temporary issue (e.g. captive portal redirecting requests, some badly configured network, no cellular connectivity, some issue with the upstream server).

Implement this after #2727 is done.

**Comment by kkaefer** [comment]
This is folded into #2727.
